internal/action/main_test.go

```diff
@@ -0,0 +1,291 @@
+package action
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/go-tfe"
+	"github.com/stretchr/testify/assert"
+)
+
+var workspacePrefix string = "action-test"
+
+func NewTestRunConfig() *RunConfig {
+	return &RunConfig{
+		Token:                  os.Getenv("tf_token"),
+		Organization:           os.Getenv("tf_organization"),
+		Host:                   "app.terraform.io",
+		Name:                   fmt.Sprintf("%s-%d", workspacePrefix, time.Now().Unix()),
+		Import:                 true,
+		Apply:                  true,
+		TFEProviderVersion:     "0.25.3",
+		RunnerTerraformVersion: "1.0.5",
+		TerraformVersion:       "1.0.5",
+	}
+}
+
+func RemoveTestWorkspaces(ctx context.Context, client *tfe.Client, organization string, prefix string) error {
+	workspaces, err := client.Workspaces.List(ctx, organization, tfe.WorkspaceListOptions{
+		Search: &prefix,
+		ListOptions: tfe.ListOptions{
+			PageSize: 100,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, ws := range workspaces.Items {
+		if err := client.Workspaces.DeleteByID(ctx, ws.ID); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func TestCreateWorkspace(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+
+	ctx := context.Background()
+
+	config := NewTestRunConfig()
+
+	client, err := tfe.NewClient(&tfe.Config{
+		Address: fmt.Sprintf("https://%s", config.Host),
+		Token:   config.Token,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := RemoveTestWorkspaces(ctx, client, config.Organization, workspacePrefix); err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = client.Workspaces.Read(ctx, config.Organization, config.Name)
+	if err == nil {
+		t.Fatal("workspace should not exist, and an error should be returned")
+	}
+
+	if err.Error() != "resource not found" {
+		t.Fatalf("Error is not workspace not found: %s", err)
+	}
+
+	Run(config)
+
+	ws, err := client.Workspaces.Read(ctx, config.Organization, config.Name)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, ws.Name, config.Name)
+
+	t.Cleanup(func() {
```

Comment:
Cleanup hooks should be registered as soon as possible. Cleanup will never run if `Read` errors. Cleanups should always be registered _before_ the stateful work happens.